cmd/sample-generic-controlplane/server/server.go

```diff
@@ -0,0 +1,189 @@
+				return utilerrors.NewAggregate(errs)
+			}
+
+			// add feature enablement metrics
+			utilfeature.DefaultMutableFeatureGate.AddMetrics()
+			ctx := genericapiserver.SetupSignalContext()
+			return Run(ctx, completedOptions)
+		},
+		Args: func(cmd *cobra.Command, args []string) error {
+			for _, arg := range args {
+				if len(arg) > 0 {
+					return fmt.Errorf("%q does not take any arguments, got %q", cmd.CommandPath(), args)
+				}
+			}
+			return nil
+		},
+	}
+
+	var namedFlagSets cliflag.NamedFlagSets
+	s.AddFlags(&namedFlagSets)
+	verflag.AddFlags(namedFlagSets.FlagSet("global"))
+	globalflag.AddGlobalFlags(namedFlagSets.FlagSet("global"), cmd.Name(), logs.SkipLoggingConfigurationFlags())
+
+	fs := cmd.Flags()
+	for _, f := range namedFlagSets.FlagSets {
+		fs.AddFlagSet(f)
+	}
+
+	cols, _, _ := term.TerminalSize(cmd.OutOrStdout())
+	cliflag.SetUsageAndHelpFunc(cmd, namedFlagSets, cols)
+
+	return cmd
+}
+
+// Run runs the specified APIServer. This should never exit.
+func Run(ctx context.Context, opts options.CompletedOptions) error {
+	// To help debugging, immediately log version
+	klog.Infof("Version: %+v", version.Get())
+
+	klog.InfoS("Golang settings", "GOGC", os.Getenv("GOGC"), "GOMAXPROCS", os.Getenv("GOMAXPROCS"), "GOTRACEBACK", os.Getenv("GOTRACEBACK"))
+
+	config, err := NewConfig(opts)
+	if err != nil {
+		return err
+	}
+	completed, err := config.Complete()
+	if err != nil {
+		return err
+	}
+	server, err := CreateServerChain(completed)
+	if err != nil {
+		return err
+	}
+
+	prepared, err := server.PrepareRun()
+	if err != nil {
+		return err
+	}
+
+	return prepared.Run(ctx)
+}
+
+// CreateServerChain creates the apiservers connected via delegation.
+func CreateServerChain(config CompletedConfig) (*aggregatorapiserver.APIAggregator, error) {
+	// 1. CRDs
+	notFoundHandler := notfoundhandler.New(config.ControlPlane.Generic.Serializer, genericapifilters.NoMuxAndDiscoveryIncompleteKey)
+	apiExtensionsServer, err := config.APIExtensions.New(genericapiserver.NewEmptyDelegateWithCustomHandler(notFoundHandler))
+	if err != nil {
+		return nil, fmt.Errorf("failed to create apiextensions-apiserver: %w", err)
+	}
+	crdAPIEnabled := config.APIExtensions.GenericConfig.MergedResourceConfig.ResourceEnabled(apiextensionsv1.SchemeGroupVersion.WithResource("customresourcedefinitions"))
+
+	// 2. Natively implemented resources
+	nativeAPIs, err := config.ControlPlane.New("sample-generic-controlplane", apiExtensionsServer.GenericAPIServer)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create generic controlplane apiserver: %w", err)
+	}
+	client, err := kubernetes.NewForConfig(config.ControlPlane.Generic.LoopbackClientConfig)
+	if err != nil {
+		return nil, err
+	}
+	storageProviders, err := config.ControlPlane.GenericStorageProviders(client.Discovery())
+	if err != nil {
+		return nil, fmt.Errorf("failed to create storage providers: %w", err)
+	}
+	if err := nativeAPIs.InstallAPIs(storageProviders...); err != nil {
```

Comment:
would we like to add some example of how to add native types to this binary as well, or is that for later/already covered with the sample-apiserver based off k8s.io/apiserver?

Reply:
See the minimal example. That's meant to have only a few native types.